File: 백준/Silver/2531. 회전 초밥/회전 초밥.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.HashSet;
import java.util.Set;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int D = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());
        int C = Integer.parseInt(st.nextToken());
        int[] dish = new int[N];
        int[] sushi = new int[D + 1];
        int result = 0;
        int max = 0;

        for(int i = 0; i < N; i++) {
            dish[i] = Integer.parseInt(br.readLine());
            if(i < K) {
                if(sushi[dish[i]] == 0) {
                    max++;
                }
                sushi[dish[i]]++;
            }
        }
        result = max;
        for(int i = 0; i < N; i++) {
            int prev = dish[i];
            int next = dish[(i + K) % N];
            if(sushi[prev] == 1) {
                max--;
            }
            sushi[prev]--;

            if(sushi[next] == 0) {
                max++;
            }
            sushi[next]++;
            if(result <= max) {
                result = max;
                if(sushi[C] == 0) {
                    result++;
                }
            }
        }

        System.out.println(result);
    }
}

File: 백준/Silver/2531. 회전 초밥/회전 초밥.kt
import java.util.StringTokenizer
import kotlin.math.max

//8 30 4 30
//7
//9
//7
//30
//2
//7
//9
//25
fun main() = with(System.`in`.bufferedReader()) {
    val st = StringTokenizer(readLine())
    val N = st.nextToken().toInt()
    val sushi = st.nextToken().toInt()
    val K = st.nextToken().toInt()
    val bonus = st.nextToken().toInt()
    var kind = 0

    val arr = IntArray(N)
    val dish = IntArray(sushi+1)

    repeat(N) {
        arr[it] = readLine().toInt()
        if(it < K) {
            if(dish[arr[it]] == 0) kind++
            dish[arr[it]]++
        }
    }


    var answer = if(dish[bonus] == 0) kind + 1 else kind

    repeat(N) {
        dish[arr[it]]--
        if(dish[arr[it]] == 0) kind--

        val end = (it + K) % N
        if(dish[arr[end]] == 0) kind++
        dish[arr[end]]++

        answer = if(dish[bonus] == 0) {
            max(answer, kind + 1)
        } else {
            max(answer, kind)
        }
    }

    print(answer)
}